File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    
    def bfs(begin,target):
        q = deque()
        v = set()
        
        q.append((begin,0))
        v.add(begin)
        
        while q:
            next,cnt = q.popleft()

            if next==target:
                return cnt
            
            for i in range(len(words)):
                if words[i] in v:   continue
                diff_cnt = 0
                for j in range(len(words[i])):
                    if words[i][j]!=next[j]:
                        diff_cnt+=1
                #print(diff_cnt,words[i],next)
                if diff_cnt==1:
                    q.append((words[i],cnt+1))
                    v.add(words[i])
                    cnt+=1
                    
        return 0

    if target not in words:
        answer = 0
    else:
        answer = bfs(begin,target)
    
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    static class Point {
        String word;
        int cnt;
        
        Point(String word, int cnt) {
            this.word = word;
            this.cnt = cnt;
        }
    }
    
    static int bfs(String begin, String target, String[] words) {
        Queue<Point> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        
        q.offer(new Point(begin, 0));
        visited.add(begin);
        
        while (!q.isEmpty()) {
            Point cur = q.poll();
            if (cur.word.equals(target)) return cur.cnt;
            
            for (String nxt: words) {
                if (visited.contains(nxt)) continue;
                int num = 0;
                for (int i=0; i<cur.word.length(); i++) {
                    if (cur.word.charAt(i)!=nxt.charAt(i)) num++;
                }
                if (num!=1) continue;
                q.offer(new Point(nxt, cur.cnt+1));
                visited.add(nxt);
            }
        }
        
        
        return -1;
        
    }
    
    public int solution(String begin, String target, String[] words) {
        if (!Arrays.asList(words).contains(target)) return 0;
        
        int answer = bfs(begin, target, words);
        return answer;
    }
}